text/3529-cargo-path-bases.md

```diff
@@ -0,0 +1,426 @@
+that exists is found. However, this introduces additional possibilities
+for user confusion if, say, `foo` exists in multiple paths in
+`CARGO_PATHS` and the first one is picked (though maybe that could be a
+warning?).
+
+NodeJS (with npm) is very similar to Python, except that dependencies
+can also be
+[specified](https://nodejs.org/api/modules.html#modules_all_together)
+using relative paths like Cargo's `path` dependencies. For non-path
+dependencies, it searches in [`node_modules/` in every parent
+directory](https://nodejs.org/api/modules.html#modules_loading_from_node_modules_folders),
+as well as in the [`NODE_PATH` search
+path](https://nodejs.org/api/modules.html#modules_loading_from_the_global_folders).
+There does not exist a standard mechanism to specify a path dependency
+relative to a path named elsewhere. With CommonJS modules, JavaScript
+developers are able to extrapolate variables directly into their
+`require` arguments, and can thus implement custom schemes for getting
+customizable paths.
+
+Ruby's `Gemfile` [path
+dependencies](https://bundler.io/man/gemfile.5.html#PATH) are only ever
+absolute paths or paths relative to the `Gemfile`'s location, and so are
+similar to Rust's current `path` dependencies.
+
+The same is the case for Go's `go.mod` [replacement
+dependencies](https://golang.org/doc/modules/managing-dependencies#tmp_10),
+which only allow absolute or relative paths.
+
+From this, it's clear that other major languages do not have a feature
+quite like this. This is likely because path dependencies are assumed
+to be short-lived and local, and thus having them be host-specific is
+often good enough. However, as the motivation section of this RFC
+outlines, there are still use-cases where a simple name-indirection
+could help.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+## Path bases relative to other path bases
+
+We could allow defining a path base relative to another path base:
+
+```toml
+[path-bases]
+base1 = "/dev/me"
+base2 = { base = "base1", path = "some_subdir" } # /dev/me/some_subdir
+```
+
+## Path dependency with just a base
+
+We could allow defining a path dependency with *just* `base`, making
+`cratename = { base = "thebase" }` equivalent to
+`cratename = { base = "thebase", path = "cratename" }`. This would simplify many
+common cases, where crates appear within the base in a directory named for the
+crate.
+
+## Git dependencies
+
+It seems reasonable to extend path bases to `git` dependencies, with something
+like:
+
+```toml
+[path-bases]
+gh = "https://github.com/jonhoo"
+```
+
+```toml
+[dependency]
+foo = { git = "foo.git", base = "gh" }
+```
+
+However, this may get complicated if someone specifies `git`, `path`, _and_
+`base`.
+
+## Support patches
+
+It may also be useful to be able to use path bases in `patch` and `path` tables.

```

Comment:
With how intertwined dependencies and patches are in Cargo, this should probably be called out in the main section of the RFC that patches are not intended to be supported.

Reply:
We talked about this in today's Cargo meeting and lean towards going ahead and supporting patches now.
- In theory, we'll get support for nearly free.  We'll need to test various corner cases and *maybe* something will be uncovered that needs development
- Otherwise, we'd have to add explicit error checks and tests for the various corner cases to ensure the error check happens

Reply:
Ok, pulled the `patch` table into scope